File: lib/sentiero/redaction.rb
# frozen_string_literal: true

require "uri"
require_relative "redaction/config"

module Sentiero
  # Redaction engine for side-channel capture (navigation, form_submit,
  # metadata, error, click) that bypasses rrweb input masking. Must stay
  # byte-for-byte equivalent to the JS twin frontend/src/redaction.js;
  # test/fixtures/redaction_cases.json pins that parity.
  module Redaction
    REDACTED = "[redacted]"

    # Fixed application order, identical to the JS module.
    TEXT_PATTERN_ORDER = %i[url jwt email long_hex card].freeze

    TEXT_PATTERNS = {
      jwt: /eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/,
      email: /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/,
      long_hex: /\b[0-9a-fA-F]{32,}\b/,
      card: /\b\d(?:[ -]?\d){12,18}\b/
    }.freeze

    URL_IN_TEXT = %r{https?://\S+}

    BUILTIN_DENYLIST = %w[
      token access_token refresh_token id_token password passwd pwd secret
      api_key apikey key sig signature code auth session sessionid otp
    ].freeze

    CUSTOM_EVENT_TYPE = 5
    META_EVENT_TYPE = 4

    URL_METADATA_KEYS = %w[url referrer entry_url entry_referrer].freeze

    # tag => { "field" => :url|:text }
    CUSTOM_FIELD_MAP = {
      "navigation" => {"url" => :url, "text" => :text},
      "__form_submit" => {"url" => :url},
      "error" => {"message" => :text, "stack" => :text, "source" => :url},
      "__click" => {"selector" => :text}
    }.freeze

    module_function

    def redact_url(url, config = Config.new)
      return url unless url.is_a?(String)

      case config.url_mode
      when :keep_all then url
      when :keep_filtered then filter_url(url, config)
      else strip_url_string(url)
      end
    end

    def redact_text(value, config = Config.new)
      return value unless value.is_a?(String)

      out = value
      config.active_text_patterns.each do |name|
        out = apply_text_pattern(out, name)
      end
      config.custom_patterns.each { |re| out = out.gsub(re, REDACTED) }
      out
    end

    def redact_event(event, config = Config.new)
      return event unless event.is_a?(Hash)

      if event["type"] == CUSTOM_EVENT_TYPE && event["data"].is_a?(Hash)
        redact_custom_event(event, config)
      elsif event["type"] == META_EVENT_TYPE && event["data"].is_a?(Hash)
        redact_meta_event(event, config)
      else
        redact_dom_event(event, config)
      end
    end

    def redact_metadata(metadata, config = Config.new)
      return metadata unless metadata.is_a?(Hash)

      metadata.to_h do |key, value|
        if URL_METADATA_KEYS.include?(key)
          [key, redact_url(value, config)]
        else
          [key, deep_redact_strings(value, config)]
        end
      end
    end

    def apply_text_pattern(text, name)
      if name == :url
        text.gsub(URL_IN_TEXT) { |m| strip_url_string(m) }
      else
        text.gsub(TEXT_PATTERNS.fetch(name), REDACTED)
      end
    end

    def filter_url(url, config)
      base, query, frag = split_url(url)
      pairs = query.empty? ? [] : query.split("&").filter_map { |p| filter_param(p, config) }
      out = base
      out += "?#{pairs.join("&")}" unless pairs.empty?
      out += "##{redact_text(frag, config)}" unless frag.empty?
      out
    end

    def filter_param(pair, config)
      eq = pair.index("=")
      name = (eq ? pair[0...eq] : pair).downcase
      # Denylist wins over the allowlist so allowlisting a built-in secret name
      # (token/password/...) can't re-enable persisting it.
      return nil if config.effective_denylist.include?(name)
      return pair if config.effective_allowlist.include?(name)
      return pair unless eq

      # Match patterns against the decoded value (email=user%40example.com must
      # be caught the same as email=user@example.com) but only substitute when
      # something actually matched; a clean survivor keeps its original,
      # unmodified encoding rather than being needlessly re-encoded.
      raw_value = pair[(eq + 1)..]
      decoded = url_decode(raw_value)
      redacted = redact_text(decoded, config)
      (redacted == decoded) ? pair : "#{pair[0...eq]}=#{redacted}"
    end

    # Plain percent-decode (leaves "+" alone, unlike www-form decoding). Falls
    # back to the raw value on malformed escapes or invalid UTF-8 rather than
    # raising, since this parses attacker-controlled URLs from public events.
    def url_decode(value)
      decoded = URI::RFC2396_PARSER.unescape(value)
      decoded.valid_encoding? ? decoded : value
    end

    # Manual split (not URI) so JS and Ruby behave identically on edge cases.
    def split_url(url)
      base = url
      frag = ""
      if (h = base.index("#"))
        frag = base[(h + 1)..]
        base = base[0...h]
      end
      query = ""
      if (q = base.index("?"))
        query = base[(q + 1)..]
        base = base[0...q]
      end
      [base, query, frag]
    end

    def strip_url_string(url)
      cut = url.index("?") || url.length
      hash = url.index("#")
      cut = hash if hash && hash < cut
      url[0...cut]
    end

    def redact_custom_event(event, config)
      map = CUSTOM_FIELD_MAP[event["data"]["tag"]]
      payload = event["data"]["payload"]
      return event unless payload.is_a?(Hash)

      # Mapped fields use their url/text treatment; every other field (and every
      # field of an unmapped tag) is deep-redacted rather than stored raw, so a
      # buggy/hostile client can't smuggle PII through an unmapped key.
      new_payload = payload.to_h do |k, v|
        case map&.dig(k)
        when :url then [k, redact_url(v, config)]
        when :text then [k, redact_text(v, config)]
        else [k, deep_redact_strings(v, config)]
        end
      end
      event.merge("data" => event["data"].merge("payload" => new_payload))
    end

    # DOM text/data is left alone unless the operator opts in.
    def redact_dom_event(event, config)
      return event if config.dom_patterns.empty? && config.custom_patterns.empty?
      return event unless event["data"]

      dom_cfg = Config.new(disabled_patterns: TEXT_PATTERN_ORDER - config.dom_patterns,
        custom_patterns: config.custom_patterns)
      event.merge("data" => deep_redact_strings(event["data"], dom_cfg))
    end

    # rrweb Meta events (type 4) carry the full page URL in data.href, which
    # bypasses rrweb's own input masking entirely; always URL-redact it like
    # any other structural URL field (navigation.url, error.source, ...).
    def redact_meta_event(event, config)
      return event unless event["data"].key?("href")

      event.merge("data" => event["data"].merge("href" => redact_url(event["data"]["href"], config)))
    end

    def deep_redact_strings(value, config)
      case value
      when String then redact_text(value, config)
      when Array then value.map { |v| deep_redact_strings(v, config) }
      when Hash
        # Keys can carry PII too (e.g. a caller using an email as a hash key);
        # redact them the same as values. Last-write-wins on key collisions.
        value.to_h { |k, v| [deep_redact_strings(k, config), deep_redact_strings(v, config)] }
      else value
      end
    end
  end
end

File: frontend/src/redaction.js
// Shared redaction engine, twin of lib/sentiero/redaction.rb. Patterns and
// application order MUST match; test/fixtures/redaction_cases.json pins parity.

export const REDACTED = "[redacted]";
export const TEXT_PATTERN_ORDER = ["url", "jwt", "email", "long_hex", "card"];

const TEXT_PATTERNS = {
  jwt: /eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g,
  email: /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g,
  long_hex: /\b[0-9a-fA-F]{32,}\b/g,
  card: /\b\d(?:[ -]?\d){12,18}\b/g,
};
const URL_IN_TEXT = /https?:\/\/\S+/g;

// Built-in secret param names always dropped in keepFiltered mode. Must match
// BUILTIN_DENYLIST in lib/sentiero/redaction.rb.
export const BUILTIN_DENYLIST = [
  "token", "access_token", "refresh_token", "id_token", "password", "passwd", "pwd", "secret",
  "api_key", "apikey", "key", "sig", "signature", "code", "auth", "session", "sessionid", "otp",
];

export function parseConfig(raw = {}) {
  return {
    urlMode: raw.urlMode || "strip",
    allowlist: (raw.urlParamAllowlist || []).map((s) => s.toLowerCase()),
    denylist: [...BUILTIN_DENYLIST, ...(raw.urlParamDenylist || []).map((s) => s.toLowerCase())],
    disabled: raw.disabledPatterns || [],
    // A pattern that is valid in Ruby but not JS must not throw out of init and
    // disable recording; skip the bad one and keep the rest.
    customRegexes: (raw.customPatterns || []).flatMap((s) => {
      try {
        return [new RegExp(s, "g")];
      } catch {
        console.warn(`[Sentiero] skipping invalid custom redaction pattern: ${s}`);
        return [];
      }
    }),
  };
}

function stripUrlString(url) {
  let cut = url.indexOf("?");
  const hash = url.indexOf("#");
  if (cut < 0) cut = url.length;
  if (hash >= 0 && hash < cut) cut = hash;
  return url.slice(0, cut);
}

function applyTextPattern(text, name) {
  if (name === "url") return text.replace(URL_IN_TEXT, (m) => stripUrlString(m));
  return text.replace(TEXT_PATTERNS[name], REDACTED);
}

export function redactText(value, cfg) {
  if (typeof value !== "string") return value;
  let out = value;
  for (const name of TEXT_PATTERN_ORDER) {
    if (!cfg.disabled.includes(name)) out = applyTextPattern(out, name);
  }
  for (const re of cfg.customRegexes) out = out.replace(re, REDACTED);
  return out;
}

function splitUrl(url) {
  let base = url;
  let frag = "";
  const h = base.indexOf("#");
  if (h >= 0) { frag = base.slice(h + 1); base = base.slice(0, h); }
  let query = "";
  const q = base.indexOf("?");
  if (q >= 0) { query = base.slice(q + 1); base = base.slice(0, q); }
  return { base, query, frag };
}

// Plain percent-decode (leaves "+" alone, unlike www-form decoding). Falls
// back to the raw value on a malformed escape rather than throwing, since
// this parses attacker-controlled URLs from public events.
function urlDecode(value) {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
}

function filterParam(pair, cfg) {
  const eq = pair.indexOf("=");
  const name = (eq >= 0 ? pair.slice(0, eq) : pair).toLowerCase();
  // Denylist wins over the allowlist so allowlisting a built-in secret name
  // (token/password/...) can't re-enable persisting it.
  if (cfg.denylist.includes(name)) return null;
  if (cfg.allowlist.includes(name)) return pair;
  if (eq < 0) return pair;

  // Match patterns against the decoded value (email=user%40example.com must
  // be caught the same as email=user@example.com) but only substitute when
  // something actually matched; a clean survivor keeps its original,
  // unmodified encoding rather than being needlessly re-encoded.
  const rawValue = pair.slice(eq + 1);
  const decoded = urlDecode(rawValue);
  const redacted = redactText(decoded, cfg);
  return redacted === decoded ? pair : pair.slice(0, eq) + "=" + redacted;
}

export function redactUrl(url, cfg) {
  if (typeof url !== "string") return url;
  if (cfg.urlMode === "keepAll") return url;
  if (cfg.urlMode !== "keepFiltered") return stripUrlString(url);

  const { base, query, frag } = splitUrl(url);
  const pairs = query ? query.split("&").map((p) => filterParam(p, cfg)).filter((p) => p !== null) : [];
  let out = base;
  if (pairs.length) out += "?" + pairs.join("&");
  if (frag) out += "#" + redactText(frag, cfg);
  return out;
}

// Structural layer: which field of which side-channel event/metadata is a URL vs
// free text. MUST match lib/sentiero/redaction.rb (CUSTOM_FIELD_MAP / URL_METADATA_KEYS);
// the custom_event/metadata cases in redaction_cases.json pin that parity.
export const CUSTOM_EVENT_TYPE = 5;
export const META_EVENT_TYPE = 4;
export const URL_METADATA_KEYS = ["url", "referrer", "entry_url", "entry_referrer"];
export const CUSTOM_FIELD_MAP = {
  navigation: { url: "url", text: "text" },
  __form_submit: { url: "url" },
  error: { message: "text", stack: "text", source: "url" },
  __click: { selector: "text" },
};

export function deepRedactStrings(value, cfg) {
  if (typeof value === "string") return redactText(value, cfg);
  if (Array.isArray(value)) return value.map((v) => deepRedactStrings(v, cfg));
  if (value && typeof value === "object") {
    const out = {};
    // Keys can carry PII too (e.g. a caller using an email as an object key);
    // redact them the same as values. Last-write-wins on key collisions.
    for (const [k, v] of Object.entries(value)) out[deepRedactStrings(k, cfg)] = deepRedactStrings(v, cfg);
    return out;
  }
  return value;
}

export function redactPayload(tag, payload, cfg) {
  if (payload == null || typeof payload !== "object") return payload;
  const map = CUSTOM_FIELD_MAP[tag];
  const out = {};
  // Mapped fields use their url/text treatment; every other field (and every
  // field of an unmapped tag) is deep-redacted rather than passed through raw.
  for (const [k, v] of Object.entries(payload)) {
    if (map && map[k] === "url") out[k] = redactUrl(v, cfg);
    else if (map && map[k] === "text") out[k] = redactText(v, cfg);
    else out[k] = deepRedactStrings(v, cfg);
  }
  return out;
}

export function redactEvent(event, cfg) {
  if (event && event.type === CUSTOM_EVENT_TYPE && event.data && typeof event.data === "object") {
    return { ...event, data: { ...event.data, payload: redactPayload(event.data.tag, event.data.payload, cfg) } };
  }
  // rrweb Meta events carry the full page URL in data.href, which bypasses
  // rrweb's own input masking entirely; always URL-redact it like any other
  // structural URL field (navigation.url, error.source, ...).
  if (event && event.type === META_EVENT_TYPE && event.data && typeof event.data === "object" && "href" in event.data) {
    return { ...event, data: { ...event.data, href: redactUrl(event.data.href, cfg) } };
  }
  return event;
}

export function redactMetadata(metadata, cfg) {
  if (metadata == null || typeof metadata !== "object") return metadata;
  const out = {};
  for (const [k, v] of Object.entries(metadata)) {
    out[k] = URL_METADATA_KEYS.includes(k) ? redactUrl(v, cfg) : deepRedactStrings(v, cfg);
  }
  return out;
}
